aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -0,0 +1,87 @@
+#include <ATen/ATen.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/native/cuda/ForeachUtils.cuh>
+
+namespace at { namespace native {
+
+namespace {
+
+// TensorListMetadata has to be < 4KB - the limit for kernel launch argument
+static constexpr int depth_to_max_tensors[5] = {110, 64, 48, 36, 30};
+static constexpr int depth_to_max_blocks[5] = {320, 320, 320, 320, 320};
+
+template<int n> struct TensorListMetadata
+{
+  void* addresses[n][depth_to_max_tensors[n-1]];
+  int sizes[depth_to_max_tensors[n-1]];
+  unsigned char block_to_tensor[depth_to_max_blocks[n-1]];
+  int block_to_chunk[depth_to_max_blocks[n-1]];
+};
+
+template<typename T, typename U, typename... ArgTypes>
+C10_LAUNCH_BOUNDS_1(kBlockSize)
+__global__ void 
+multi_tensor_apply_kernel(
+    T tensorListMeta,
+    U callable,
+    ArgTypes... args) {
+  // Hand the chunk information to the user-supplied functor to process however it likes.
+  callable(kChunkSize, tensorListMeta, args...); 
+}
+
+template<int depth, typename T, typename... ArgTypes>
+void multi_tensor_apply(
+    std::vector<std::vector<at::Tensor>>& tensor_lists,
+    T callable,
+    ArgTypes... args) {
+        TORCH_CHECK(tensor_lists.size() == depth, "Amount of tensor lists has to match the depth.");
+
+        int n_tensors = tensor_lists[0].size();
+        TensorListMetadata<depth> tensorListMeta;
+
```

Comment:
per comment in native_functions, put OptionalCUDAGuard set on the device of one of the tensor arguments here:
```
const at::cuda::OptionalCUDAGuard device_guard(device_of(tensor_lists[0][0]));
```

Reply:
Done